domain/upgrade/state/state.go

```diff
@@ -0,0 +1,310 @@
+WHERE  upgrade_info_controller_node.upgrade_info_uuid = $M.info_uuid`
+	s, err := sqlair.Prepare(q, infoControllerNode{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var nodeInfos []infoControllerNode
+	err = tx.Query(ctx, s, sqlair.M{"info_uuid": upgradeUUID}).GetAll(&nodeInfos)
+	if err != nil && err != sqlair.ErrNoRows {
+		return nil, errors.Trace(err)
+	}
+	return nodeInfos, nil
+}
+
+// CreateUpgrade creates an active upgrade to and from specified versions
+// and returns the upgrade's UUID
+func (st *State) CreateUpgrade(ctx context.Context, previousVersion, targetVersion version.Number) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	upgradeUUID := utils.MustNewUUID().String()
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		q := `
+INSERT INTO upgrade_info (uuid, previous_version, target_version, created_at) VALUES
+($M.uuid, $M.previous_version, $M.target_version, DATETIME('now'))`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":             upgradeUUID,
+			"previous_version": previousVersion.String(),
+			"target_version":   targetVersion.String(),
+		}).Run())
+	})
+
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return upgradeUUID, nil
+}
+
+// SetControllerReady marks the supplied controllerID as being ready
+// to start a provided upgrade. All provisioned controllers need to
+// be ready before an upgrade can start
+// A controller node is ready for an upgrade if a row corresponding
+// to the controller is present in upgrade_info_controller_node
+func (st *State) SetControllerReady(ctx context.Context, upgradeUUID, controllerID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		nodeInfos, err := st.getUpgradeNodes(ctx, tx, upgradeUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, nodeinfo := range nodeInfos {
+			if nodeinfo.ControllerNodeID == controllerID {
+				return nil
+			}
+		}
+		q := `
+INSERT INTO upgrade_info_controller_node (uuid, controller_node_id, upgrade_info_uuid) VALUES
+($M.uuid, $M.controller_id, $M.info_uuid)`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":          utils.MustNewUUID().String(),
+			"controller_id": controllerID,
+			"info_uuid":     upgradeUUID,
+		}).Run())
+	}))
+}
+
+// AllProvisionedControllersReady returns true if and only if all controllers
+// that have been started by the provisioner are ready to start the provided upgrade
+func (st *State) AllProvisionedControllersReady(ctx context.Context, upgradeUUID string) (bool, error) {
```

Comment:
You don't need to be pulling records out of the DB to do this. 

The relational algebra has the answer, if you can form the request. Something like this:
```sql
SELECT NULL
FROM   controller_node_id n
       LEFT JOIN upgrade_info_controller_node u
       ON n.controller_id = u.controller_node_id
WHERE  u.upgrade_info_uuid = ?  
AND    n.dqlite_node_id IS NOT NULL
AND    u.controller_node_id IS NULL
```
If it returns any rows, all the controllers are not yet ready.

Incidentally, I would rename controller_node_id to controller_id in upgrade_info_controller_node.